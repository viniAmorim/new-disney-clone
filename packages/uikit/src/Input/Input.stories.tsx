import React from 'react';

import { Story, Meta } from '@storybook/react';

import Input, { InputProps } from '~/Input';

export default {
  title: 'Form/Input',
  component: Input,
} as Meta;

const Template: Story<InputProps> = (args) => <Input {...args} />;

export const Default = Template.bind({});
Default.args = {
  placeholder: 'Text content for input',
};

export const Error = Template.bind({});
Error.args = {
  placeholder: 'Text content for input',
  isValid: false,
};

export const Mask = Template.bind({});
Mask.args = {
  placeholder: 'Text content for input',
  format: '(##) ####-####',
  mask: '_',
};
