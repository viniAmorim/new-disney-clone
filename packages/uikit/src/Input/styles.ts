import { NumberFormatProps } from 'react-number-format';

// semantic error TS4023: Exported variable
import NumberFormat from '../../../../node_modules/react-number-format/dist/react-number-format';

import styled, { css } from 'styled-components';

export interface StyledInputProps extends NumberFormatProps {
  isValid?: boolean;
}

const variantModifier = {
  default: css`
    border: 1px solid ${({ theme }) => theme.colors['gray-400']};
  `,
  success: css`
    border: 1px solid ${({ theme }) => theme.colors.success};
  `,
  error: css`
    border: 1px solid ${({ theme }) => theme.colors.error};
  `,
};

const InputCss = css`
  background: ${({ theme }) => theme.colors.white};
  border: 1px solid ${({ theme }) => theme.colors['gray-400']};
  border-radius: 4px;
  padding: 12px 16px;

  &:focus,
  &:hover {
    border: 1px solid ${({ theme }) => theme.colors['blue-300']};
    outline: none;

    ::placeholder {
      color: ${({ theme }) => theme.colors['gray-100']};
    }
  }
`;

export const Input = styled(NumberFormat).withConfig({
  shouldForwardProp: (prop, defaultValidatorFn) => {
    return !['isValid'].includes(prop as string) && defaultValidatorFn(prop);
  },
})<StyledInputProps>`
  ${InputCss}

  ${({ isValid }) =>
    isValid ? variantModifier.default : variantModifier.error};
`;

export const RegularInput = styled.input.withConfig({
  shouldForwardProp: (prop, defaultValidatorFn) =>
    !['isValid'].includes(prop) && defaultValidatorFn(prop),
})<StyledInputProps>`
  ${InputCss}

  ${({ isValid }) =>
    isValid ? variantModifier.default : variantModifier.error};
`;

export const InputError = styled.text`
  margin-top: 6px;
  color: ${({ theme }) => theme.colors.error};
`;
