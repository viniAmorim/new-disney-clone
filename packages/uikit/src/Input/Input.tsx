import React, { forwardRef } from 'react';

import * as Styled from './styles';

export interface InputProps extends Styled.StyledInputProps {
  /**
   * Placeholder
   */
  placeholder: string;
}

const Input = forwardRef(function({ isValid, mask, ...rest }: InputProps, ref) {
  if (mask) {
    return <Styled.Input ref={ref} isValid={isValid} {...rest} />;
  }

  return <Styled.RegularInput ref={ref} isValid={isValid} {...rest} />;
});

export default Input;
