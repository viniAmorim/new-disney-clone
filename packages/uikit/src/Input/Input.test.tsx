/* eslint-disable @typescript-eslint/no-var-requires */
import React from 'react';

import { act, fireEvent, render } from '~/test/utils';

import Input from './Input';

describe('Input', () => {
  it('Should render the component', () => {
    const { getByTestId } = render(
      <Input data-testid="input" placeholder="Text content for input" />,
    );

    const textFieldElement = getByTestId('input');

    expect(textFieldElement).toBeInTheDocument();
    expect(textFieldElement).toMatchSnapshot();
  });

  it('Should render a input error variant', () => {
    const { getByRole } = render(
      <Input isValid={false} placeholder="Text content for input" />,
    );

    const inputElement = getByRole('textbox');

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toMatchSnapshot();
  });

  it('can have focus', () => {
    const { getByRole } = render(
      <Input placeholder="Text content for input" />,
    );

    const inputElement = getByRole('textbox');

    expect(inputElement).toBeInTheDocument();

    act(() => {
      fireEvent.keyDown(document.body, { key: 'TAB' });
      inputElement.focus();
    });

    expect(inputElement).toHaveFocus();

    expect(inputElement).toMatchSnapshot();
  });
});
