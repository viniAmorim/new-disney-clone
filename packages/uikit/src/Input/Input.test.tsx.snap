// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Should render a input error variant 1`] = `
<input
  class="sc-gsTEea dIyyZv"
  placeholder="Text content for input"
/>
`;

exports[`Input Should render the component 1`] = `
<input
  class="sc-gsTEea dIyyZv"
  data-testid="input"
  placeholder="Text content for input"
/>
`;

exports[`Input can have focus 1`] = `
<input
  class="sc-gsTEea dIyyZv"
  placeholder="Text content for input"
/>
`;
